src/zcash/IncrementalMerkleTree.cpp

```diff
@@ -0,0 +1,297 @@
+#include <stdexcept>
+
+#include <boost/foreach.hpp>
+
+#include "zcash/IncrementalMerkleTree.hpp"
+#include "crypto/sha256.h"
+#include "zerocash/utils/util.h" // TODO: remove these utilities
+
+namespace libzcash {
+
+SHA256Compress SHA256Compress::combine(const SHA256Compress& a, const SHA256Compress& b)
+{
+    // This is a performance optimization.
+    if (a.IsNull() && b.IsNull()) {
+        return a;
+    }
+
+    SHA256Compress res = SHA256Compress();
+
+    CSHA256 hasher;
+    hasher.Write(a.begin(), 32);
+    hasher.Write(b.begin(), 32);
+    hasher.FinalizeNoPadding(res.begin());
+
+    return res;
+}
+
+
+template <typename Hash>
+class PathFiller {
+private:
+    std::deque<Hash> queue;
+public:
+    PathFiller() : queue() { }
+    PathFiller(std::deque<Hash> queue) : queue(queue) { }
+
+    Hash next() {
+        if (queue.size() > 0) {
+            Hash h = queue.front();
+            queue.pop_front();
+
+            return h;
+        } else {
+            return Hash();
+        }
+    }
+};
+
+template<size_t Depth, typename Hash>
+void IncrementalMerkleTree<Depth, Hash>::wfcheck() const {
```

Comment:
Those two trees would not be equivalent, but you can't deserialize (2) with the API with my proposed wfcheck. (prevent `left` from being empty when *either* `right` or `parents` are not)

> you don't really need right at all, since you can always hash it into left right after you set it

If you're trying to witness `right`, `left` is part of the authentication path, [see here.](https://github.com/zcash/zcash/pull/823/commits/3a32400b876ddadd60d3b9ee300031d8fd6774c0#diff-0f8d8f2984419bfb2f6f55897878ff66R202)